### PR TITLE
Run plot_labels in different process to prevent main process crash

### LINKED ISSUE #17307: Training seg models hangs then errors at Plotting labels step
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

Config parameters used
```
    'model_name': 'yolo11s-seg.pt',
    'epochs': 50,
    'imgsz': 800,
    'optimizer': 'SGD',
    'pretrained': False,
    'batch': -1,
    'workers': 6,
    'mosaic': 0.0,
    'flipud': 0.5,
    'fliplr': 0.5,
    'scale': 0.1,
    'translate': 0.05,
```

Ultralytics prints:
```
Ultralytics 8.3.27 🚀 Python-3.10.12 torch-2.5.1+cu124 CUDA:0 (NVIDIA A100-SXM4-40GB, 40339MiB)
[34m[1mengine/trainer: [0mtask=segment, mode=train, model=yolo11s-seg.pt, data=/home/guillaume/tc/remote-inspection/unity_datasets/config/bale_v2.yaml, epochs=50, time=None, patience=25, batch=-1, imgsz=800, save=True, save_period=-1, cache=False, device=None, workers=6, project=/home/guillaume/tc/remote-inspection/trained_models/bale_v2_yolo11s-seg, name=train, exist_ok=False, pretrained=False, optimizer=SGD, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=None, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=True, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.05, scale=0.1, shear=0.0, perspective=0.0, flipud=0.5, fliplr=0.5, bgr=0.0, mosaic=0.0, mixup=0.0, copy_paste=0.0, copy_paste_mode=flip, auto_augment=randaugment, erasing=0.4, crop_fraction=1.0, cfg=None, tracker=botsort.yaml, save_dir=/home/guillaume/tc/remote-inspection/trained_models/bale_v2_yolo11s-seg/train
Overriding model.yaml nc=80 with nc=1

                   from  n    params  module                                       arguments                     
  0                  -1  1       928  ultralytics.nn.modules.conv.Conv             [3, 32, 3, 2]                 
  1                  -1  1     18560  ultralytics.nn.modules.conv.Conv             [32, 64, 3, 2]                
  2                  -1  1     26080  ultralytics.nn.modules.block.C3k2            [64, 128, 1, False, 0.25]     
  3                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]              
  4                  -1  1    103360  ultralytics.nn.modules.block.C3k2            [128, 256, 1, False, 0.25]    
  5                  -1  1    590336  ultralytics.nn.modules.conv.Conv             [256, 256, 3, 2]              
  6                  -1  1    346112  ultralytics.nn.modules.block.C3k2            [256, 256, 1, True]           
  7                  -1  1   1180672  ultralytics.nn.modules.conv.Conv             [256, 512, 3, 2]              
  8                  -1  1   1380352  ultralytics.nn.modules.block.C3k2            [512, 512, 1, True]           
  9                  -1  1    656896  ultralytics.nn.modules.block.SPPF            [512, 512, 5]                 
 10                  -1  1    990976  ultralytics.nn.modules.block.C2PSA           [512, 512, 1]                 
 11                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 12             [-1, 6]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 13                  -1  1    443776  ultralytics.nn.modules.block.C3k2            [768, 256, 1, False]          
 14                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 15             [-1, 4]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 16                  -1  1    127680  ultralytics.nn.modules.block.C3k2            [512, 128, 1, False]          
 17                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]              
 18            [-1, 13]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 19                  -1  1    345472  ultralytics.nn.modules.block.C3k2            [384, 256, 1, False]          
 20                  -1  1    590336  ultralytics.nn.modules.conv.Conv             [256, 256, 3, 2]              
 21            [-1, 10]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 22                  -1  1   1511424  ultralytics.nn.modules.block.C3k2            [768, 512, 1, True]           
 23        [16, 19, 22]  1   1474291  ultralytics.nn.modules.head.Segment          [1, 32, 128, [128, 256, 512]] 
YOLO11s-seg summary: 355 layers, 10,082,675 parameters, 10,082,659 gradients, 35.6 GFLOPs

Transferred 555/561 items from pretrained weights
Freezing layer 'model.23.dfl.conv.weight'
[34m[1mAMP: [0mrunning Automatic Mixed Precision (AMP) checks...
Downloading https://github.com/ultralytics/assets/releases/download/v8.3.0/yolo11n.pt to 'yolo11n.pt'...

  0%|          | 0.00/5.35M [00:00<?, ?B/s]
100%|██████████| 5.35M/5.35M [00:00<00:00, 193MB/s]
[34m[1mAMP: [0mchecks passed ✅
[34m[1mAutoBatch: [0mComputing optimal batch size for imgsz=800 at 60.0% CUDA memory utilization.
[34m[1mAutoBatch: [0mCUDA:0 (NVIDIA A100-SXM4-40GB) 39.39G total, 0.14G reserved, 0.12G allocated, 39.13G free
      Params      GFLOPs  GPU_mem (GB)  forward (ms) backward (ms)                   input                  output
    10082675       55.61         0.568         47.08           nan        (1, 3, 800, 800)                    list
    10082675       111.2         1.401         37.87           nan        (2, 3, 800, 800)                    list
    10082675       222.5         2.649         38.35           nan        (4, 3, 800, 800)                    list
    10082675       444.9         5.125         41.66           nan        (8, 3, 800, 800)                    list
    10082675       889.8        10.110         46.23           nan       (16, 3, 800, 800)                    list
    10082675        1780        19.973         79.05           nan       (32, 3, 800, 800)                    list
    10082675        3559        39.642         154.3           nan       (64, 3, 800, 800)                    list
[34m[1mAutoBatch: [0mUsing batch-size 37 for CUDA:0 23.27G/39.39G (59%) ✅

```

Training then hangs for ~1h and terminates with a `Killed`

```
New cache created: []/val/labels.cache
Plotting labels to []/bale_v2_yolo11s-seg/train/labels.jpg...
```

### Environment

```
Ultralytics 8.3.27 🚀 Python-3.10.12 torch-2.5.1+cu124 CUDA:0 (NVIDIA A100-SXM4-40GB, 40339MiB)
Setup complete ✅ (12 CPUs, 83.5 GB RAM, 69.3/193.6 GB disk)

OS                  Linux-6.8.0-1015-gcp-x86_64-with-glibc2.35
Environment         Linux
Python              3.10.12
Install             pip
RAM                 83.46 GB
Disk                69.3/193.6 GB
CPU                 Intel Xeon 2.20GHz
CPU count           12
GPU                 NVIDIA A100-SXM4-40GB, 40339MiB
GPU count           1
CUDA                12.4

numpy               ✅ 1.26.4>=1.23.0
matplotlib          ✅ 3.9.2>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 10.4.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.14.1>=1.4.1
torch               ✅ 2.5.1>=1.8.0
torchvision         ✅ 0.20.1>=0.9.0
tqdm                ✅ 4.66.6>=4.64.0
psutil              ✅ 5.9.8
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.10>=2.0.0
numpy               ✅ 1.26.4<2.0.0; sys_platform == "darwin"
torch               ✅ 2.5.1!=2.4.0,>=1.8.0; sys_platform == "win32"

```

### Minimal Reproducible Example

This was trained using custom data but nothing seemed out of the ordinary. Training works with Ultralytics 8.3.20  

### Additional

Seems related to known issue https://github.com/ultralytics/yolov5/issues/5395

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Y-T-G**
Closes https://github.com/ultralytics/ultralytics/issues/17307

Currently, `TryExcept` is used to catch plotting crash in https://github.com/ultralytics/yolov5/issues/5395. 
https://github.com/ultralytics/ultralytics/blob/6806f15396432fffb951250d650454b840eb4c28/ultralytics/utils/plotting.py#L859-L861

But since that's still using the main process, if it crashes, it can't be caught and the main process would die. The workaround is to use multi-processing so that the plotting runs in a different process altogether and doesn't kill the main process when it crashes.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Introduce a decorator to manage long-running functions by running them in a separate process with an optional timeout mechanism.

### 📊 Key Changes
- Added a new `forked` decorator function to `ultralytics/utils/__init__.py`:
  - Allows functions to run in a separate process.
  - Supports specifying a timeout to automatically terminate functions taking too long.
- Applied the `forked` decorator to `plot_labels` in `ultralytics/utils/plotting.py` to address known execution time issues.

### 🎯 Purpose & Impact
- **Purpose**: Enhance the stability and reliability of the code by preventing long-running functions from blocking operations with a configurable timeout feature.
- **Impact**: Users can expect more robust function execution, particularly preventing indefinite hangs when plotting labels, which improves performance and user experience.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting a pull request to the `ultralytics/ultralytics` repository! 🚀 Your contribution is greatly appreciated. Here’s a checklist to ensure everything is on track for a smooth integration:

- ✅ **Define a Purpose**: Please make sure the purpose of your fix is clearly explained in your PR description. You have linked the related issue, which is great! Ensure that your commit messages are clear and concise, following the project's conventions.
- ✅ **Synchronize with Source**: Confirm that your PR is up to date with the `ultralytics/ultralytics` `main` branch. If it's not, please update it by clicking the 'Update branch' button or using `git pull` and `git merge main` on your local setup.
- ✅ **Ensure CI Checks Pass**: Verify that all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address these issues promptly.
- ✅ **Update Documentation**: If your changes involve new features or modifications, please update the relevant [documentation](https://docs.ultralytics.com).
- ✅ **Add Tests**: Where applicable, include or update tests to cover your changes and verify that all tests are passing.
- ✅ **Sign the CLA**: Since this is your first Ultralytics pull request, make sure you have signed the [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/). You can do so by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Aim to keep your changes to the **minimum** required for your bug fix or enhancement. _"It is not daily increase, but daily decrease, hack away the unessential."_ — Bruce Lee

🔍 For any bug fixes, it would be helpful to provide a minimum reproducible example (MRE) if you haven't already, to assist in diagnosing the issue effectively.

For more detailed instructions, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). If you have any questions or need further assistance, feel free to leave a comment here. An Ultralytics engineer will be with you shortly to provide further support. Thank you for contributing to Ultralytics! 🎉
